File: 0781-rabbits-in-forest/hashmap/cpp/solution.cpp
class Solution {
public:
    int numRabbits(vector<int>& answers) {
        unordered_map<int, int> ans_freq_map;
        for (int ans : answers) {
            ans_freq_map[ans]++;
        }
        int total_rabbits = 0;
        for (auto const& [ans, freq] : ans_freq_map) {
            int group_size = ans + 1;
            int num_groups = ceil(double(freq) / group_size); 
            total_rabbits += num_groups * group_size;
        }
        return total_rabbits;
    }
};

File: 0781-rabbits-in-forest/hashmap/java/solution.java
class Solution {
    public int numRabbits(int[] answers) {
        Map<Integer, Integer> ansFreqMap = new HashMap<>();
        for (int ans : answers) {
            ansFreqMap.put(ans, ansFreqMap.getOrDefault(ans, 0) + 1);
        }

        int totalRabbits = 0;
        for (var entry : ansFreqMap.entrySet()) {
            int ans = entry.getKey();
            int freq = entry.getValue();
            int groupSize = ans + 1;
            int numGroups = (int) Math.ceil((double) freq / groupSize);
            totalRabbits += numGroups * groupSize;
        }
        return totalRabbits;
    }
}
